src/lib.rs

```diff
@@ -220,47 +227,46 @@ struct StatementData {
     params: u16,
 }
 
-impl<B: MysqlShim<W>, R: Read, W: Write> MysqlIntermediary<B, R, W> {
-    /// Create a new server over two one-way channels and process client commands until the client
+impl<B: MysqlShim<RW>, RW: Read + Write> MysqlIntermediary<B, RW> {
+    /// Create a new server over a two-way channel and process client commands until the client
     /// disconnects or an error occurs.
-    pub fn run_on(shim: B, reader: R, writer: W) -> Result<(), B::Error> {
-        let r = packet::PacketReader::new(reader);
-        let w = packet::PacketWriter::new(writer);
-        let mut mi = MysqlIntermediary {
-            shim,
-            reader: r,
-            writer: w,
-        };
+    pub fn run_on(shim: B, rw: RW) -> Result<(), B::Error> {
+        let rw = packet::PacketConn::new(rw);
+        let mut mi = MysqlIntermediary { shim, rw };
         mi.init()?;
         mi.run()
     }
 
     fn init(&mut self) -> Result<(), B::Error> {
-        self.writer.write_all(&[10])?; // protocol 10
+        self.rw.write_all(&[10])?; // protocol 10
 
         // 5.1.10 because that's what Ruby's ActiveRecord requires
-        self.writer.write_all(&b"5.1.10-alpha-msql-proxy\0"[..])?;
-
-        self.writer.write_all(&[0x08, 0x00, 0x00, 0x00])?; // TODO: connection ID
-        self.writer.write_all(&b";X,po_k}\0"[..])?; // auth seed
-        self.writer.write_all(&[0x00, 0x42])?; // just 4.1 proto
-        self.writer.write_all(&[0x21])?; // UTF8_GENERAL_CI
-        self.writer.write_all(&[0x00, 0x00])?; // status flags
-        self.writer.write_all(&[0x00, 0x00])?; // extended capabilities
-        self.writer.write_all(&[0x00])?; // no plugins
-        self.writer.write_all(&[0x00; 6][..])?; // filler
-        self.writer.write_all(&[0x00; 4][..])?; // filler
-        self.writer.write_all(&b">o6^Wz!/kM}N\0"[..])?; // 4.1+ servers must extend salt
-        self.writer.flush()?;
+        self.rw.write_all(&b"5.1.10-alpha-msql-proxy\0"[..])?;
+
+        self.rw.write_all(&[0x08, 0x00, 0x00, 0x00])?; // TODO: connection ID
+        self.rw.write_all(&b";X,po_k}\0"[..])?; // auth seed
+        let capabilities = &mut [0x00, 0x42]; // 4.1 proto
+        if self.shim.tls_config().is_some() {
```

Comment:
It seems possibly unfortunate that we're calling this multiple times, sometimes throwing away the result. I wonder if we should just call `.tls_config()` in `::new` and then save it so we can then use it in places like here. What do you think?

Reply:
I agree. Done & pushed.